src/sql/src/pure.rs

```diff
@@ -478,24 +479,87 @@ async fn purify_format_single(
-                    let csv_header = file.lines().next_line().await?;
-                    match csv_header {
-                        Some(csv_header) => {
-                            csv_header
-                                .split(*delimiter as char)
-                                .for_each(|v| col_names.push(Ident::from(v)));
+            let first_row = if let Some(file) = file {
+                let file = tokio::io::BufReader::new(file);
+                let csv_header = file.lines().next_line().await;
+                match csv_header {
+                    Ok(Some(csv_header)) => Some(
+                        csv_header
+                            .split(*delimiter as char)
+                            .map(|name| name.to_string())
+                            .collect::<Vec<_>>(),
+                    ),
+                    Ok(None) => {
+                        if matches!(columns, CsvColumns::Header) {
+                            bail!(
+                                "CSV file expected to have at least one line \
+                                 to determine column names, but is empty"
+                            );
+                        } else {
+                            None
                         }
-                        None => bail!("CSV file expected header line, but is empty"),
                     }
-                } else {
-                    bail!("CSV format with headers only works with file connectors")
+                    Err(e) => {
+                        // TODO(#7562): support compressed files
+                        if matches!(columns, CsvColumns::Header) {
+                            bail!("Cannot determine header by reading CSV file: {}", e);
+                        } else {
+                            None
+                        }
+                    }
+                }
+            } else {
+                None
+            };
+
+            match (&columns, first_row) {
+                (CsvColumns::Header, Some(cols)) => {
+                    *columns = CsvColumns::HeaderNamed(cols.into_iter().map(Ident::from).collect())
+                }
+                (CsvColumns::Count(n), first_line) => {
+                    if let Some(columns) = first_line {
+                        if *n != columns.len() {
+                            bail!(
+                                "Specified column count (WITH {} COLUMNS) \
+                                 does not match number of columns in CSV file ({})",
+                                n,
+                                columns.len()
+                            );
+                        }
+                    }
+
+                    if col_names.is_empty() {
+                        *col_names = (1..=*n)
+                            .into_iter()
+                            .map(|i| Ident::from(format!("column{}", i)))
+                            .collect();
+                    }
```

Comment:
Isn't this the thing we don't want to do anymore? Like, the thing that couples the source alias to the format specifier?

Reply:
Yeah, that was why I proposed the syntax that you objected to in the design doc (with column names and header-ness being orthogonal concepts).

Without dedicated syntax, this is the only thing available to us to record the schema in the catalog. We can avoid doing this and continue just relying on [`DataEncoding::desc()`](https://github.com/MaterializeInc/materialize/blob/2c5dada709e1dc5/src/dataflow-types/src/types.rs#L529-L533) but I thought that we agreed that having the schema in the catalog was worth doing.

If we don't want to record it, then just deleting this code is fine, everything should continue to work.